explore/marginal-gain/1.Marginal-Gain-Multiple-Mutations.py

```diff
@@ -0,0 +1,259 @@
+    series['testing_auroc'] = roc_auc_score(y_test, y_pred_test)
+    series['n_pos_coeffs'] = n_pos
+    series['n_neg_coeffs'] = n_neg
+    series['n_positive_mutation'] = sum(y==1)
+    series['n_negative_mutation'] = sum(y==0)
+    series['cum_rank_cov_feat'] = cov_ranks.sum()
+    series['median_rank_cov_feat'] = np.median(cov_ranks)
+    series['mean_rank_cov_feat'] = np.mean(cov_ranks)
+    return series
+
+def grid_scores_to_df(grid_scores):
+    """
+    Convert a sklearn.grid_search.GridSearchCV.grid_scores_ attribute to 
+    a tidy pandas DataFrame where each row is a hyperparameter-fold combinatination.
+    """
+    rows = list()
+    for grid_score in grid_scores:
+        for fold, score in enumerate(grid_score.cv_validation_scores):
+            row = grid_score.parameters.copy()
+            row['fold'] = fold
+            row['score'] = score
+            rows.append(row)
+    df = pd.DataFrame(rows)
+    return df
+
+def get_coeffs(pipeline, X_train, model_type):
+    """
+    Get a dataframe with the training coefficients 
+    """
+    final_pipeline = pipeline.best_estimator_
+    final_classifier = final_pipeline.named_steps['classify']
+    
+    # Get indices of features
+    if model_type == 'model a': 
+        select_indices = list(range(len(covariates.columns.values)))
+    else:
+        select_indices = final_pipeline.named_steps['features'].transform(
+            np.arange(len(X_train.columns)).reshape(1, -1)
+        ).tolist()
+        select_indices = [x for sublist in select_indices for x in sublist]
+    
+    # Make df features, weights
+    coef_df = pd.DataFrame.from_items([
+        ('feature', X_train.columns[select_indices]),
+        ('weight', final_classifier.coef_[0]),
+    ])
+
+    return coef_df
+
+def get_sign_coeffs(coef_df):
+    return (coef_df.weight>0).sum(), (coef_df.weight<0).sum()
+
+def get_ranks_covariates_feat(coef_df):
+    coef_df['abs'] = coef_df['weight'].abs()
+    coef_df = coef_df.sort_values('abs', ascending=False)
+    
+    def RepresentsInt(s):
+        try: 
+            int(s)
+            return True
+        except ValueError:
+            return False
+
+    coef_df['is_cov_feat'] = [not RepresentsInt(x) for x in coef_df['feature']]
+    ranks = np.flatnonzero(coef_df['is_cov_feat'])
+    return ranks
+
+auroc_dfs = {}
+
+
+# ## Train the models.
+
+# In[12]:
+
+get_ipython().run_cell_magic('time', '', '# Train model a: covariates only.\nwarnings.filterwarnings("ignore") # ignore deprecation warning for grid_scores_\nrows = list()\nfor m in list(mutations):\n    series = pd.Series()\n    series[\'mutation\'] = m\n    series[\'symbol\'] = mutations[m]\n    rows.append(get_aurocs(X[\'model a\'], Y[m], cv_pipeline[\'model a\'], series, \'model a\'))\nauroc_dfs[\'model a\'] = pd.DataFrame(rows)\nauroc_dfs[\'model a\'].sort_values([\'symbol\', \'testing_auroc\'], ascending=[True, False], inplace=True)\ndisplay(auroc_dfs[\'model a\'])')
```

Comment:
Let's remove the `%%time` magic. While very helpful, I think line by line commenting on the diff here will be more usefull.